### PR TITLE
(PUP-9465) Add libuser provider to group type

### PR DESCRIPTION
**Pull request by GabrielNagy**
Currently, group provider `groupadd` uses two separate packages for managing groups. The (groupadd, groupmod, groupdel) programs belong to the shadow project. These utilities explicitly operate on local files (/etc/passwd, /etc/shadow, /etc/group, /etc/gshadow) only.

The (lgroupadd, lgroupmod, lgroupdel) programs belong to the libuser project. Per its documentation, libuser is a library that implements a standardized interface for manipulating and administering user and group accounts, using pluggable back-ends to interface to data sources.

To fix this mis-engineering, we separate the two providers and deprecate the `forcelocal` parameter which is only used in conjunction with the `libuser` commands. We also make use of libuser's ability of directly managing group members in order to benefit from Puppet's `manages_members` feature.

However, we avoid defaulting the libuser provider since it does not have the same set of features as the groupadd provider, hence it should be an opt-in provider. For example, libuser doesn't support NIS accounts.

Still needs proper tests.

### CODE REVIEW

**Comment by GabrielNagy** [comment]
jenkins please test this on redhat7-64a

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa,amazon6-64a,amazon7-64a,debian8-64a,debian8-32a,debian9-64a,debian9-32a,debian10-64a,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat7-POWERa,redhat7-AARCH64a,redhat8-64a,redhatfips7-64a,sles11-64a,sles11-32a,sles12-64a,sles12-POWERa,sles15-64a,solaris10-64a,solaris11-64a,solaris114-64a,ubuntu1404-64a,ubuntu1404-32a,ubuntu1604-64a,ubuntu1604-32a,ubuntu1604-POWERa,ubuntu1804-64a,windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2016-64a,windows2019-64a

**Comment by GabrielNagy** [comment]
Some lore behind the changes:

This PR (as is) does _not_ change current Puppet functionality, apart from the fact that it defaults the `groupadd` provider to Linux machines. This is being done because currently `groupadd` is the single suitable group provider for Linux systems. This causes a warning to show up when the new provider is introduced, because both are considered suitable and none is defaulted. Since we don't want to default the new one, we explicitly default `groupadd`.

Deprecation messages were also added to the `forcelocal` parameter of the existing provider should we want to remove it in a future release (in favor of the `libuser` provider).

Apart from this, if an user wants to use the new provider, they have to explicitly mention it in the manifest.

Another feature which was not present in the `groupadd` provider, but was added in `libuser` is the ability to manage members from a group resource, i.e.:

```puppet
group { 'testgroup':
  ensure     => present,
  name       => 'testgroup',
  members    => ['user3', 'user1', 'user2'],
  provider   => 'libuser',
}
```

**Comment by mihaibuzgau** [comment]
jenkins please test this on aix61-POWERfa,aix71-POWERfa,aix72-POWERfa,amazon6-64a,amazon7-64a,debian8-64a,debian8-32a,debian9-64a,debian9-32a,debian10-64a,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat7-POWERa,redhat7-AARCH64a,redhat8-64a,redhatfips7-64a,sles11-64a,sles11-32a,sles12-64a,sles12-POWERa,sles15-64a,solaris10-64a,solaris11-64a,solaris114-64a,ubuntu1404-64a,ubuntu1404-32a,ubuntu1604-64a,ubuntu1604-32a,ubuntu1604-POWERa,ubuntu1804-64a,windows10ent-64a,windows10ent-32a,windows2008r2-64a,windows2012r2-64a,windows2016-64a,windows2019-64a

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,,ubuntu1404-64a,ubuntu1404-32a,ubuntu1604-64a,windows10ent-64a,windows10ent-32a

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,,ubuntu1404-64a,ubuntu1404-32a,ubuntu1604-64a,windows10ent-64a,windows10ent-32a

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,ubuntu1604-64a,windows10ent-64a,windows10ent-32a

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,ubuntu1604-64a,windows10ent-64a,windows10ent-32a

File: lib/puppet/provider/group/libuser.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    cmd << @resource[:name]
+    cmd
+  end
+
+  def deletecmd
+    [command(:delete), @resource[:name]]
+  end
+
+  def members_insync?(current, should)
+    current.sort == @resource.parameter(:members).actual_should(current, should)
```
Comment (ciprianbadescu) [inline lib/puppet/provider/group/libuser.rb]: should `@resource.parameter(:members).actual_should(current, should)` also be sorted?

File: lib/puppet/type/group.rb [hunk 2]
```diff
@@ -96,7 +96,7 @@ def change_to_s(currentvalue, newvalue)
 
       def insync?(current)
         if provider.respond_to?(:members_insync?)
-          return provider.members_insync?(current, @should)
+          return provider.members_insync?(current, munge_members_value(@should))
```
Comment (ciprianbadescu) [inline lib/puppet/type/group.rb]: not sure about this modification, as generic behavior gets changes, could the change be done in libuser.rb?

File: lib/puppet/provider/group/libuser.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  def members_to_s(current)
+    return '' if current.nil? or !current.kind_of?(Array)
+    current.join(',')
+  end
+
+  def member_valid?(user)
+    begin
+      !!Etc.getpwnam(user)
+    rescue ArgumentError
+      raise ArgumentError, _("User %{user} does not exist") % { user: user }
```
Comment (GabrielNagy) [inline lib/puppet/provider/group/libuser.rb]: should we raise a Puppet error instead of an ArgumentError?

File: lib/puppet/provider/group/libuser.rb (hunk 1 above)
Comment (GabrielNagy) [reply lib/puppet/provider/group/libuser.rb]: it's sorted as per here https://github.com/puppetlabs/puppet/blob/77fe57b788e10341c16af8d5f21823784aa23f4f/lib/puppet/type/group.rb#L130, but we can have a test for it

File: lib/puppet/type/group.rb (hunk 2 above)
Comment (GabrielNagy) [reply lib/puppet/type/group.rb]: yes, I can implement this in `libuser#members_insync?`

**Comment by GabrielNagy** [comment]
@joshcooper do you happen to know whether we should accept comma separated values for the `members` parameter of the `group` resource?

i.e. should we handle/munge `members => "user1,user2,user3"` or should we only care about `members => ['user1', 'user2', 'user3']`

File: lib/puppet/provider/group/groupadd.rb [hunk 4]
```diff
@@ -69,6 +72,7 @@ def addcmd
     if @resource.forcelocal?
       cmd = [command(:localadd)]
       @custom_environment = Puppet::Util::Libuser.getenv
+      Puppet.deprecation_warning _("`forcelocal` is deprecated and will be removed in a future release. This property has been replaced by the `libuser` provider.")
```
[the remark this answers is not in the document]
Comment (GabrielNagy) [reply lib/puppet/provider/group/groupadd.rb]: yeah, I would create a `libuser` provider for the user type as well.

[the remark this answers is not in the document]
Comment (melissa) [reply lib/puppet/provider/group/groupadd.rb]: Are you going to do that as a part of this PR or in a separate PR?

[the remark this answers is not in the document]
Comment (GabrielNagy) [reply lib/puppet/provider/group/groupadd.rb]: I filed PUP-10168 for the user part.

**Comment by joshcooper** [comment]
Kicking Travis

**Comment by mihaibuzgau** [comment]
jenkins please test this on aix61-POWERfa,fedora28-64a,fedora29-64a,fedora30-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,ubuntu1604-64a,windows10ent-64a,windows10ent-32a

**Comment by GabrielNagy** [comment]
I thought about this for a bit, and I believe this provider separation doesn't bring much improvement apart from the code being cleaner.

With https://github.com/puppetlabs/puppet/pull/7936 being in the works, it wouldn't make sense for us to deprecate `forcelocal`, since AIX can also benefit from that parameter.

What we can do though, is further drive away the misconception that the L in the `luser`/`lgroup` commands stands for `local`, when actually it's because of `libuser`.

PUP-10169 keeps the `libuser` feature, but only uses it to conditionally activate the new `manages_local_users_and_groups` feature. Using the same thought process, we can do the same for `manages_members` (which is the only functional addition this PR made) and move its implementation to `groupadd.rb`.

tl;dr drop this PR, implement the libuser-specific `manages_members` feature in the existing `groupadd` provider...

**Comment by GabrielNagy** [comment]
Closing in favor of https://github.com/puppetlabs/puppet/pull/7953
